src/esp/sim/Simulator.cpp

```diff
@@ -1125,7 +1125,7 @@ agent::Agent::ptr Simulator::getAgent(const int agentId) {
 }
 
 esp::sensor::Sensor::ptr Simulator::addSensorToObject(
-    const int objectId,
+    int objectId,
     esp::sensor::SensorSpec::ptr& sensorSpec) {
```

Comment:
Oh, just saw your detailed error msg in the afternoon.
This is because one cannot bind a non-const lvalue reference to an rvalue of type.

`esp::sensor::SensorSpec::ptr&` is a non-const reference of type `esp::sensor::SensorSpec::ptr`. However, when you call this function from SimTest, the argument type is actually `esp::sensor::CameraSensorSpec::ptr` which is being converted to `esp::SensorSpec::ptr` so that a temporary, (a rvalue) is created (which you did not realize). 

Solution is to add a `const`, namely, `const esp::sensor::SensorSpec::ptr&`.

Please try it and see if it works.

Reply:
Thank you that works and makes a lot of sense! I understand more about this now